File: javascript/packages/ethereum/src/PrivateDataPayloadValidator.js
import { ethers } from 'ethers';
import { createAttestationSuccess, createAttestationFailure } from '../../base/src/AttestationVerifier.js';
import { AttestationReasonCodes } from '../../base/src/AttestationReasonCodes.js';
import { validateMerkleRootMatch } from './MerkleRootValidator.js';

/**
 * Validates payloads for the PrivateData schema.
 * For PrivateData, the attestation data is a raw 32-byte Merkle root hash.
 * Validation succeeds if the data equals the expected Merkle root.
 */
class PrivateDataPayloadValidator {
  /**
   * Creates a new PrivateDataPayloadValidator.
   * @param {Object} [logger] - Optional logger for diagnostic information
   */
  constructor(logger = null) {
    this.logger = logger;
  }

  /**
   * Validates that the attestation data encodes the expected Merkle root.
   * For PrivateData schema, attestation data should be exactly 32 bytes matching the expected Merkle root.
   *
   * @param {string | Uint8Array} attestationData - Raw attestation data from the on-chain attestation
   * @param {string} expectedMerkleRoot - Expected Merkle root value (as hex string, with or without 0x prefix)
   * @param {string} attestationUid - UID of the attestation being validated (for error reporting)
   * @returns {Promise<Object>} AttestationResult with isValid, message, reasonCode, and attestationUid
   */
  async validatePayloadAsync(attestationData, expectedMerkleRoot, attestationUid) {
    // Use centralized validator (see docs/MERKLE_ROOT_BINDING.md)
    const { isValid, reasonCode } = validateMerkleRootMatch(attestationData, expectedMerkleRoot);

    if (isValid) {
      this.logger?.log?.('debug', `PrivateData payload validation successful for attestation ${attestationUid}`);
      return createAttestationSuccess(
        'PrivateData payload matches expected Merkle root',
        attestationUid,
        AttestationReasonCodes.VALID
      );
    }

    let message;
    let reasonCodeEnum;

    if (reasonCode === 'INVALID_ATTESTATION_DATA') {
      message = 'PrivateData attestation data is null or empty';
      reasonCodeEnum = AttestationReasonCodes.INVALID_ATTESTATION_DATA;
    } else {
      const actualHex = attestationData ? ethers.hexlify(attestationData) : 'null';
      const expectedHex = expectedMerkleRoot ? ethers.hexlify(expectedMerkleRoot) : 'null';
      message = `PrivateData Merkle root mismatch. Expected: ${expectedHex}, Actual: ${actualHex}`;
      reasonCodeEnum = AttestationReasonCodes.MERKLE_MISMATCH;
    }

    this.logger?.log?.('warn', `PrivateData payload validation failed: ${message}`);
    return createAttestationFailure(message, reasonCodeEnum, attestationUid);
  }
}

export { PrivateDataPayloadValidator };

File: dotnet/src/Zipwire.ProofPack.Ethereum/ProofPack/PrivateDataPayloadValidator.cs
using Evoq.Blockchain;
using Microsoft.Extensions.Logging;
using Zipwire.ProofPack;

namespace Zipwire.ProofPack.Ethereum;

/// <summary>
/// Validates payloads for the PrivateData schema.
/// For PrivateData, the attestation data is a raw 32-byte Merkle root hash.
/// Validation succeeds if the data equals the expected Merkle root.
/// </summary>
public class PrivateDataPayloadValidator : ISchemaPayloadValidator
{
    private readonly ILogger<PrivateDataPayloadValidator>? logger;

    /// <summary>
    /// Creates a new PrivateDataPayloadValidator.
    /// </summary>
    /// <param name="logger">Optional logger for diagnostic information.</param>
    public PrivateDataPayloadValidator(ILogger<PrivateDataPayloadValidator>? logger = null)
    {
        this.logger = logger;
    }

    /// <inheritdoc />
    public Task<AttestationResult> ValidatePayloadAsync(byte[] attestationData, Hex expectedMerkleRoot, string attestationUid)
    {
        if (attestationData == null || attestationData.Length == 0)
        {
            logger?.LogWarning("PrivateData payload validation failed: attestation data is null or empty for attestation {AttestationUid}", attestationUid);
            return Task.FromResult(AttestationResult.Failure(
                "PrivateData attestation data is null or empty",
                AttestationReasonCodes.InvalidAttestationData,
                attestationUid));
        }

        // Convert attestation data to Hex for comparison
        var attestationDataHex = new Hex(attestationData);

        // Check if the attestation data equals the expected merkle root
        if (attestationDataHex.Equals(expectedMerkleRoot))
        {
            logger?.LogDebug("PrivateData payload validation successful for attestation {AttestationUid}", attestationUid);
            return Task.FromResult(AttestationResult.Success(
                "PrivateData payload matches expected Merkle root",
                string.Empty,  // attester is not available in payload validator context
                attestationUid));
        }

        logger?.LogWarning("PrivateData payload validation failed: Merkle root mismatch. Expected: {Expected}, Actual: {Actual}", expectedMerkleRoot, attestationDataHex);
        return Task.FromResult(AttestationResult.Failure(
            $"PrivateData Merkle root mismatch. Expected: {expectedMerkleRoot}, Actual: {attestationDataHex}",
            AttestationReasonCodes.MerkleMismatch,
            attestationUid));
    }
}
